client/post-editor/editor-html-toolbar/index.jsx

```diff
@@ -23,12 +34,79 @@ export class EditorHtmlToolbar extends Component {
 	};
 
 	state = {
+		isPinned: false,
+		isScrollable: false,
+		isScrolledFull: false,
 		openTags: [],
 		selectedText: '',
 		showImageDialog: false,
 		showLinkDialog: false,
 	};
 
+	componentDidMount() {
+		this.pinToolbarOnScroll = throttle( this.pinToolbarOnScroll, 50 );
+		this.hideToolbarFadeOnFullScroll = throttle( this.hideToolbarFadeOnFullScroll, 200 );
+
+		window.addEventListener( 'scroll', this.pinToolbarOnScroll );
+		window.addEventListener( 'resize', this.onWindowResize );
+		this.buttons.addEventListener( 'scroll', this.hideToolbarFadeOnFullScroll );
+
+		this.toggleToolbarScrollableOnResize();
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'scroll', this.pinToolbarOnScroll );
+		window.removeEventListener( 'resize', this.onWindowResize );
+		this.buttons.removeEventListener( 'scroll', this.hideToolbarFadeOnFullScroll );
+	}
+
+	bindButtonsRef = div => {
+		this.buttons = div;
+	}
+
+	onWindowResize = () => {
+		throttle( this.disablePinOnSmallScreens, 400 );
+		throttle( this.toggleToolbarScrollableOnResize, 200 );
+	}
+
+	pinToolbarOnScroll = () => {
+		if ( isWithinBreakpoint( '<660px' ) ) {
+			return;
+		}
+
+		const { offsetTop } = this.props.content;
+		const { isPinned } = this.state;
+
+		if ( isPinned && window.pageYOffset < offsetTop - TOOLBAR_HEIGHT ) {
+			this.setState( { isPinned: false } );
+		} else if ( ! isPinned && window.pageYOffset > offsetTop - TOOLBAR_HEIGHT ) {
+			this.setState( { isPinned: true } );
+		}
+	}
+
+	disablePinOnSmallScreens = () => {
+		if ( isWithinBreakpoint( '<660px' ) ) {
+			this.setState( { isPinned: false } );
```

Comment:
Another potential `setState` we could optimize (resizing below 660px will have frequent renders).